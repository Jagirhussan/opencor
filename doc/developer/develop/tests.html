<!DOCTYPE html>
<html>
    <head>
        <title>
            Tests
        </title>

        <meta http-equiv="content-type" content="text/html; charset=utf-8"/>

        <link href="../../3rdparty/googleCodePrettify/prettify.css" rel="stylesheet" type="text/css"/>
        <link href="../res/stylesheet.css" rel="stylesheet" type="text/css"/>

        <script src="../../3rdparty/googleCodePrettify/prettify.js" type="text/javascript"></script>
        <script src="../../3rdparty/jQuery/jquery.js" type="text/javascript"></script>
        <script src="../../res/common.js" type="text/javascript"></script>
        <script src="../res/menu.js" type="text/javascript"></script>
    </head>
    <body onload="prettyPrint()">
        <script type="text/javascript">
            headerAndContentsMenu("Tests", "../..");
        </script>

        <p>
            Tests can be created at the plugin level. Tests must:
        </p>

        <ul>
            <li>Be located under the <code>tests</code> folder (e.g. <a href="https://github.com/opencor/opencor/blob/master/src/plugins/misc/Compiler/tests"><code>[OpenCOR]/src/plugins/misc/Compiler/tests</code></a>) of a plugin folder (e.g. <a href="https://github.com/opencor/opencor/blob/master/src/plugins/misc/Compiler"><code>[OpenCOR]/src/plugins/misc/Compiler</code></a>);</li>
            <li>Consist of at least one set of <code>.cpp</code> and <code>.h</code> files (e.g. <a href="https://github.com/opencor/opencor/blob/master/src/plugins/misc/Compiler/tests/tests.cpp"><code>[OpenCOR]/src/plugins/misc/Compiler/tests/tests.cpp</code></a> and <a href="https://github.com/opencor/opencor/blob/master/src/plugins/misc/Compiler/tests/tests.h"><code>[OpenCOR]/src/plugins/misc/Compiler/tests/tests.h</code></a>);</li>
            <li>
                Be listed under the <code>TESTS</code> header in the call to our CMake <a href="https://github.com/opencor/opencor/blob/master/cmake/common.cmake"><code>ADD_PLUGIN</code></a> macro of the plugin's <code>CMakeLists.txt</code> file (e.g. <a href="https://github.com/opencor/opencor/blob/master/src/plugins/misc/Compiler/CMakeLists.txt"><code>[OpenCOR]/src/plugins/misc/Compiler/CMakeLists.txt</code></a>):
                <pre class="prettyprint">PROJECT(MyPlugin)

# Add the plugin

ADD_PLUGIN(MyPlugin
    ...
    TESTS
        tests1
        tests2
        tests3
        ...
)</pre>
            </li>
            <li>
                Be referenced in the <a href="https://github.com/opencor/opencor/blob/master/tests/main.cpp">main <code>.cpp</code> file</a> for our tests:
                <pre class="prettyprint">...

// The different tests that are  to be run

Tests tests;

...
tests["MyPlugin"] = QStringList() << "tests1" << "tests2" << "tests3" << ...;
...

...</pre>
            </li>
        </ul>

        <p>
            Some functions are available to load a plugin, run the <a href="http://en.wikipedia.org/wiki/Command-line_interface">CLI</a> version of OpenCOR and retrieve the contents of a file (see <a href="https://github.com/opencor/opencor/blob/master/tests/testsutils.cpp"><code>[OpenCOR]/tests/testsutils.cpp</code></a> and <a href="https://github.com/opencor/opencor/blob/master/tests/testsutils.h"><code>[OpenCOR]/tests/testsutils.h</code></a>).
        </p>

        <script type="text/javascript">
            copyright("../..");
        </script>
    </body>
</html>
