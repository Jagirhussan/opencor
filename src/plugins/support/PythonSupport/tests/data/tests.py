import math
import opencor as oc
import os


def test_simulation(title: str, file_name_or_url: str, first: bool = True) -> None:
    if not first:
        print()

    print('---------------------------------------')
    print(' ' * math.floor((39 - len(title)) / 2) + title)
    print('---------------------------------------')

    print(' - Open simulation...')

    if file_name_or_url.startswith('https://'):
        simulation = oc.open_simulation(file_name_or_url)
    else:
        simulation = oc.open_simulation(
            os.path.dirname(__file__) + '/../../../../../../models/tests/' + file_name_or_url)

    print(' - Check simulation...')
    print('    - Valid: ' + ('YES' if simulation.valid() else 'NO'))

    issues = simulation.issues()

    if issues:
        print('    - Issues:\n       - ' + '\n       - '.join(simulation.issues()))
    else:
        print('    - Issues: N/A')

    print(' - Run simulation...')

    simulation.run()

    print(' - Retrieve simulation results...')

    results = simulation.results()

    print(' - Retrieve states...')

    states = results.states()

    print(' - Retrieve states\' values...')

    x = states['main/x'].values()
    y = states['main/y'].values()
    z = states['main/z'].values()

    oc.close_simulation(simulation)


# Test for an unknown local/remote file

try:
    test_simulation('Unknown local file', 'unknown')
except Exception as e:
    print(repr(e))

try:
    test_simulation('Unknown remote file',
                    'https://unknown', False)
except Exception as e:
    print(repr(e))

# Test for an invalid local/remote file

try:
    test_simulation('Invalid local file', 'cellml/underconstrained_model.cellml', False)
except Exception as e:
    print(repr(e))

try:
    test_simulation('Invalid remote file',
                    'https://raw.githubusercontent.com/opencor/opencor/master/models/tests/cellml/underconstrained_model.cellml',
                    False)
except Exception as e:
    print(repr(e))

# Test for a local/remote CellML file

test_simulation('Local CellML file', 'cellml/lorenz.cellml', False)
test_simulation('Remote CellML file',
                'https://raw.githubusercontent.com/opencor/opencor/master/models/tests/cellml/lorenz.cellml', False)

# Test for a local/remote SED-ML file

test_simulation('Local SED-ML file', 'sedml/lorenz.sedml', False)
test_simulation('Remote SED-ML file',
                'https://raw.githubusercontent.com/opencor/opencor/master/models/tests/sedml/lorenz.sedml', False)

# Test for a local/remote COMBINE archive

test_simulation('Local COMBINE archive', 'combine/lorenz.omex', False)
test_simulation('Remote COMBINE archive',
                'https://raw.githubusercontent.com/opencor/opencor/master/models/tests/combine/lorenz.omex', False)
