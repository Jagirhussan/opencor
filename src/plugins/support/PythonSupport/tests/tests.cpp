/*******************************************************************************

Copyright (C) The University of Auckland

OpenCOR is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

OpenCOR is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <https://gnu.org/licenses>.

*******************************************************************************/

//==============================================================================
// Python support tests
//==============================================================================

#include "../../../../tests/src/testsutils.h"

//==============================================================================

#include "tests.h"

//==============================================================================

void Tests::basicTests()
{
    // Some very basic tests to make sure that we have access to libSBML

    QStringList output;

    QVERIFY(!OpenCOR::runCli(QStringList() << "-c" << "PythonShell" << OpenCOR::fileName("src/plugins/support/PythonSupport/tests/data/tests.py"), output));
    QCOMPARE(output, OpenCOR::fileContents(OpenCOR::fileName("src/plugins/support/PythonSupport/tests/data/tests.out")));
}

//==============================================================================

QTEST_GUILESS_MAIN(Tests)

//==============================================================================
// End of file
//==============================================================================
