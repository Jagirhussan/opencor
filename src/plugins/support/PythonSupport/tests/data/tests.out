---------------------------------------
           Local CellML file
---------------------------------------
 - Open file...
 - Run file...
 - Retrieve simulation results...
 - Retrieve states...
 - Retrieve states' values...

---------------------------------------
          Remote CellML file
---------------------------------------
 - Open file...
 - Run file...
 - Retrieve simulation results...
 - Retrieve states...
 - Retrieve states' values...

---------------------------------------
           Local SED-ML file
---------------------------------------
 - Open file...
 - Run file...
 - Retrieve simulation results...
 - Retrieve states...
 - Retrieve states' values...

---------------------------------------
          Remote SED-ML file
---------------------------------------
 - Open file...
 - Run file...
 - Retrieve simulation results...
 - Retrieve states...
 - Retrieve states' values...

---------------------------------------
         Local COMBINE archive
---------------------------------------
 - Open file...
 - Run file...
 - Retrieve simulation results...
 - Retrieve states...
 - Retrieve states' values...

---------------------------------------
        Remote COMBINE archive
---------------------------------------
 - Open file...
 - Run file...
 - Retrieve simulation results...
 - Retrieve states...
 - Retrieve states' values...
